fdbserver/DataDistribution.actor.cpp

```diff
@@ -4486,6 +4514,45 @@ ACTOR Future<Void> ddSnapCreate(DistributorSnapRequest snapReq, Reference<AsyncV
 	return Void();
 }
 
+ACTOR Future<Void> ddExclusionSafetyCheck(DistributorExclusionSafetyCheckRequest req,
+                                          Reference<DataDistributorData> self, Database cx) {
+	TraceEvent("DDExclusionSafetyCheckBegin");
+	vector<StorageServerInterface> ssis = wait(getStorageServers(cx));
+	DistributorExclusionSafetyCheckReply reply(true);
+	if (!self->teamCollection) {
+		TraceEvent("DDExclusionSafetyCheckTeamCollectionInvalid");
+		reply.safe = false;
+		req.reply.send(reply);
+		return Void();
+	}
+	vector<UID> excludeServerIDs;
+	// Go through storage server interfaces and translate Address -> server ID (UID)
+	for (const auto &ssi : ssis) {
+		for (AddressExclusion excl : req.exclusions) {
+			if (excl.excludes(ssi.address())) {
+				excludeServerIDs.push_back(ssi.id());
+				break;
+			}
+		}
+	}
+	std::sort(excludeServerIDs.begin(), excludeServerIDs.end());
+	for (const auto& team : self->teamCollection->teams) {
+		vector<UID> teamServerIDs = team->getServerIDs();
+		std::sort(teamServerIDs.begin(), teamServerIDs.end());
+		TraceEvent("DDExclusionSafetyCheck")
+			.detail("Excluding", describe(excludeServerIDs))
+			.detail("Existing", describe(teamServerIDs));
+		// If excluding set completely contains team, it is unsafe to remove these servers
+		if (std::includes(excludeServerIDs.begin(), excludeServerIDs.end(), teamServerIDs.begin(), teamServerIDs.end())) {
```

Comment:
I think it is better to make it  tunable about the minimum number of replicas a team should have in case of marking a SS as failed. For some system with more critical data, we may want at least 2 replicas in the system.

Can you make a Knob DD_EXCLUDE_MIN_REPLICAS whose default is 1? You can also BUGGIFY it to have value 2.